pkg/kubelet/oom_watcher.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/record"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/cadvisor"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	"github.com/google/cadvisor/events"
+	cadvisorApi "github.com/google/cadvisor/info/v1"
+)
+
+type OOMWatcher interface {
+	RecordSysOOMs(node *api.Node) error
+}
+
+type realOOMWatcher struct {
+	cadvisor cadvisor.Interface
+	recorder record.EventRecorder
+}
+
+func NewOOMWatcher(cadvisor cadvisor.Interface, recorder record.EventRecorder) OOMWatcher {
+	return &realOOMWatcher{
+		cadvisor: cadvisor,
+		recorder: recorder,
+	}
+}
+
+const systemOOMEvent = "SystemOOM"
+
+func (self *realOOMWatcher) RecordSysOOMs(node *api.Node) error {
+	request := events.Request{
+		EventType: map[cadvisorApi.EventType]bool{
+			cadvisorApi.EventOom: true,
+		},
+		ContainerName:        "/",
+		IncludeSubcontainers: false,
+	}
+	eventChannel, err := self.cadvisor.WatchEvents(&request)
+	if err != nil {
+		return err
+	}
+	for event := range eventChannel.GetChannel() {
+		glog.V(2).Infof("got sys oom event from cadvisor: %v", event)
+		self.recorder.PastEventf(node, util.Time{event.Timestamp}, systemOOMEvent, "System OOM encountered")
+	}
+	return fmt.Errorf("failed to watch cadvisor for sys oom events")
```

Comment:
This will happen if the connection to cAdvisor is broken. Should we just return `nil` instead of an error since it is technically OK we just need to retry?